x/datapool/keeper/pool.go

```diff
@@ -492,3 +486,121 @@ func (k Keeper) GetDataValidationCertificate(ctx sdk.Context, poolID, round uint
+		return types.ErrNFTAllIssued
+	}
+
+	if pool.GetRound() != round {
+		return types.ErrRoundNotMatched
+	}
+
+	if !payment.Equal(pool.GetPoolParams().GetNftPrice()) {
+		return types.ErrPaymentNotMatched
+	}
+
+	poolAcc, err := sdk.AccAddressFromBech32(pool.GetPoolAddress())
+	if err != nil {
+		return sdkerrors.Wrapf(err, "invalid pool address")
+	}
+
+	err = k.bankKeeper.SendCoins(ctx, buyer, poolAcc, sdk.NewCoins(payment))
+	if err != nil {
+		return sdkerrors.Wrapf(err, "failed to pay")
+	}
+
+	//mint data access NFT when pool is activated
+	if pool.GetStatus() == types.ACTIVE {
+		contractAddr := pool.GetNftContractAddr()
+
+		contractAcc, err := sdk.AccAddressFromBech32(contractAddr)
+		if err != nil {
+			return sdkerrors.Wrapf(err, "invalid NFT contract address")
+		}
+
+		mintMsg := types.NewMsgMintDataAccessNFT(pool.GetNumIssuedNfts()+1, buyer.String())
+		mintMsgBz, err := json.Marshal(mintMsg)
+		if err != nil {
+			return sdkerrors.Wrapf(err, "failed to marshal mint NFT Msg")
+		}
+
+		_, err = k.wasmKeeper.Execute(ctx, contractAcc, poolAcc, mintMsgBz, sdk.NewCoins(types.ZeroFund))
+		if err != nil {
+			return sdkerrors.Wrapf(err, "failed to mint NFT")
+		}
+	} else {
+		// if pool is PENDING state, add buyer to white list
+		k.AddToWhiteList(ctx, poolID, buyer)
+	}
+
+	k.increaseIssuedNFT(ctx, pool)
+
+	return nil
+}
+
+func (k Keeper) increaseIssuedNFT(ctx sdk.Context, pool *types.Pool) {
+	pool.NumIssuedNfts += 1
+
+	k.SetPool(ctx, pool)
+}
+
+func (k Keeper) AddToWhiteList(ctx sdk.Context, poolID uint64, addr sdk.AccAddress) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.GetKeyWhiteList(poolID, addr)
+	whiteList := &types.WhiteList{
+		PoolId:  poolID,
+		Address: addr.String(),
+	}
+
+	store.Set(key, k.cdc.MustMarshalBinaryLengthPrefixed(whiteList))
+}
+
+func (k Keeper) GetAllWhiteLists(ctx sdk.Context) ([]types.WhiteList, error) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, types.KeyPrefixPoolWhiteList)
+	defer iterator.Close()
+
+	whiteLists := make([]types.WhiteList, 0)
+
+	for ; iterator.Valid(); iterator.Next() {
+		bz := iterator.Value()
+		var list types.WhiteList
+
+		err := k.cdc.UnmarshalBinaryLengthPrefixed(bz, &list)
+		if err != nil {
+			return []types.WhiteList{}, err
+		}
+
+		whiteLists = append(whiteLists, list)
+	}
+
+	return whiteLists, nil
+}
+
+func (k Keeper) GetWhiteList(ctx sdk.Context, poolID uint64) ([]types.WhiteList, error) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, types.GetKeyPoolWhiteList(poolID))
+	defer iterator.Close()
+
+	whiteLists := make([]types.WhiteList, 0)
+
+	for ; iterator.Valid(); iterator.Next() {
+		bz := iterator.Value()
+		var list types.WhiteList
+
+		err := k.cdc.UnmarshalBinaryLengthPrefixed(bz, &list)
+		if err != nil {
+			return []types.WhiteList{}, err
+		}
+
+		whiteLists = append(whiteLists, list)
+	}
+
+	return whiteLists, nil
+}
```

Comment:
It is not a big issue, I have a question. Is we need pagination for not only `whitelist`, but also for the others?

Reply:
Good point. When I wrote the code, I forget pagination.
We may have to consider to add pagination to queries. I'll mark TODO. Thanks.